src/Data/Map/NonEmpty.hs

```diff
@@ -227,4 +227,18 @@ toNonEmpty :: NonEmptyMap k a -> NonEmpty (k, a)
 toNonEmpty (NonEmptyMap tup m) = tup :| Map.toList m
 
 toMap :: Ord k => NonEmptyMap k a -> Map.Map k a
-toMap (NonEmptyMap (k, a) m) = Map.insert k a m
\ No newline at end of file
+toMap (NonEmptyMap (k, a) m) = Map.insert k a m
+
+
+{--------------------------------------------------------------------
+  Map
+--------------------------------------------------------------------}
+
+mapWithKey :: (t -> b) -> NonEmptyMap k t -> NonEmptyMap k b
+mapWithKey f (NonEmptyMap (k, v) map) =  NonEmptyMap (k, f v) (Map.map f map)
+
+map :: (t -> b) -> NonEmptyMap k t -> NonEmptyMap k b
+map = mapWithKey
+
+mapKeys :: Ord t2 => (t -> t2) -> NonEmptyMap t a -> NonEmptyMap t2 a
```

Comment:
Since the key type is not preserved this implies that we have the ability for keys to match so this method as written has different functionality than the Map implementation due to if two values match with the same key, in Map the value of the two lesser is removed, where in this if that value is in the nonempty position than this behavior is not preserved. Now as the reduction in size only happens on key collision we know it will still be nonempty, however the function is currently not in parity with the containers implementation.

It may be clearer from the original documentation

```
O(n*log n). mapKeys f s is the map obtained by applying f to each key of s.

The size of the result may be smaller if f maps two or more distinct keys to the same new key. In this case the value at the greatest of the original keys is retained.

mapKeys (+ 1) (fromList [(5,"a"), (3,"b")])                        == fromList [(4, "b"), (6, "a")]
mapKeys (\ _ -> 1) (fromList [(1,"b"), (2,"a"), (3,"d"), (4,"c")]) == singleton 1 "c"
mapKeys (\ _ -> 3) (fromList [(1,"b"), (2,"a"), (3,"d"), (4,"c")]) == singleton 3 "c"
```

Reply:
Oh! I see! I'll work on this as soon as I have a free moment. Thanks! :relaxed: